RAGchain/benchmark/dataset/nfcorpus.py

```diff
@@ -0,0 +1,160 @@
+from copy import deepcopy
+from typing import List, Optional
+
+import pandas as pd
+
+from RAGchain.DB.base import BaseDB
+from RAGchain.benchmark.dataset.base import BaseDatasetEvaluator
+from RAGchain.pipeline.base import BaseRunPipeline
+from RAGchain.retrieval.base import BaseRetrieval
+from RAGchain.schema import EvaluateResult, Passage
+
+
+class NFCorpusEvaluator(BaseDatasetEvaluator):
+    """
+    NFCorpusEvaluator is a class for evaluating pipeline performance on NFCorpus dataset.
+    """
+
+    def __init__(self, run_pipeline: BaseRunPipeline,
+                 evaluate_size: Optional[int] = None,
+                 metrics: Optional[List[str]] = None
+                 ):
+        """
+        :param run_pipeline: The pipeline that you want to benchmark.
+        :param evaluate_size: The number of data to evaluate. If None, evaluate all data.
+        NFCorpus dataset we use is huge. Recommend to set proper size for evaluation.
+        :param metrics: The list of metrics to use. If None, use all metrics that supports NFCorpus dataset.
+        Supporting metrics are 'Recall', 'Precision', 'Hole', 'TopK_Accuracy', 'EM', 'F1_score',
+        'context_precision', 'answer_relevancy', 'faithfulness'.
+        and rank aware metrics are 'NDCG', 'AP', 'CG', 'IndDCG', 'DCG', 'IndIDCG', 'IDCG', 'RR'.
+
+        Notice:
+        The reason context_recall does not accommodate this benchmark is due to the excessive number
+        of retrieval ground truths that exceed the context length in ragas metrics.
+
+        Default metrics is basically running metrics if you run test file.
+        Support metrics is the metrics you are available.
+        This separation is because Ragas metrics take a long time in evaluation.
+        """
+
+        try:
+            import ir_datasets
+        except ImportError:
+            raise ImportError('You have to pip install ir_datasets.\n '
+                              'If it occurred error, please refer to our docs.')
+
+        file_path = "nfcorpus/test"
+        datasets = ir_datasets.load(file_path)
+
+        query = pd.DataFrame({'query_id': query[0], 'query': query[1]} for query in datasets.queries_iter())
+        doc = pd.DataFrame(
+            {'doc_id': doc[0], 'doc': doc[3], 'title': doc[2], 'url': doc[1], 'doc_metadata': datasets.docs_metadata(),
+                            'file_path': file_path} for doc in datasets.docs_iter())
+        qrels = pd.DataFrame({'query_id': qrels[0], 'retrieval_gt': qrels[1], 'relevance': qrels[2]}
+                             for qrels in datasets.qrels_iter() if qrels[2] > 1)
+
+        qrels = qrels.groupby('query_id', as_index=False).agg(
+            {'retrieval_gt': lambda x: list(x), 'relevance': lambda x: list(x)})
+
+        default_metrics = self.retrieval_gt_metrics + self.retrieval_gt_metrics_rank_aware
+        support_metrics = default_metrics + self.answer_no_gt_ragas_metrics
```

Comment:
Where is context_precision...??

Reply:
The reason context_recall does not accommodate this benchmark is due to the excessive number of retrieval ground truths that exceed the context length in ragas metrics.